extensions/sql-database-projects/src/models/project.ts

```diff
@@ -1106,6 +1113,47 @@ export class Project implements ISqlProject {
 
 		return firstPropertyElement.childNodes[0].data;
 	}
+
+	/**
+	 * Adds folder items for all intermediate folders in the path.
+	 * For example, when adding a file `a/b/c/d.sql`, we also need to add items for
+	 * the following folders:
+	 *   a/
+	 *   a/b/
+	 *   a/b/c/
+	 *
+	 * @param absolutePath Normalized absolute path to the file or folder.
+	 */
+	private async ensureFolderItemsForPath(absolutePath: string): Promise<void> {
+		const normalizedProjectFolderPath = path.normalize(this.projectFolderPath);
+
+		// Only add folders for files within the project folder. When adding files outside the project
+		// folder, they should be copied to the project root and there will be no additional folders to add.
+		if (!absolutePath.toUpperCase().startsWith(normalizedProjectFolderPath.toUpperCase())) {
+			return;
+		}
+
+		const relativePath = utils.convertSlashesForSqlProj(absolutePath.substring(normalizedProjectFolderPath.length));
+		const pathSegments = utils.trimChars(relativePath, ' \\').split('\\');
+		let intermediateFolderPath = '';
+
+		// Add folder items for all segments, except the last one, which will be explicitly
+		// added by the caller of this helper method
+		for (let segmentIndex = 0; segmentIndex < pathSegments.length - 1; ++segmentIndex) {
```

Comment:
Hm... how would you implement `length - 1` with `for ... of`? The thing is, we don't want to add the very last segment, as it will be added by the caller.

Reply:
ah ok I missed that the last one won't be added

Reply:
Maybe in conjunction with some other helper methods, yeah. I just don't see any value in introducing more operation in rather simple process. This appears to be the simplest and most performant (less operations) way to implement the required functionality.